README.md

```diff
@@ -1 +1,147 @@
+
+### What is this repository for? ###
+
+* Create new MessageClient
+* Commit to the Message Bus
+* Public messages to the Message Bus
+* Subscribe to and receives messages from the Messsage Bus
+* Disconnect from the Message Bus 
+
+### Installation ###
+
+* Make sure you have modules enabled, i.e. have an initialized  go.mod file
+* If your code is in your GOPATH then make sure ```GO111MODULE=on``` is set
+* Run ```go get github.com/edgexfoundry/go-mod-messaging```
+  * This will add the go-mod-messaging to the go.mod file and download it into the module cache
+
+### How to Use ###
+This library is used by Go programs for interacting with the Message Bus (i.e. zeroMQ). The Message Bus connection information as well as which implementation to use is stored in the service's toml configuration as:
+
+Publisher:
+```toml
+[MessageQueue]
+Protocol = 'tcp'
+Host = '*'
+Port = 5563
+Type = 'zero'
+Topic = 'events'
+```
+
+Subscriber:
+```toml
+[MessageQueue]
+Protocol = 'tcp'
+Host = 'localhost'
+Port = 5563
+Type = 'zero'
+Topic = 'events'
+```
+
+The following code snippets demonstrate how a service uses this messaging module to create a connection, send messages, and receive messages.
+
+This code snippet shows how to connect to the abstract message bus.
+
+```go
+var messageBus messaging.MessageClient
+
+var err error
+messageBus, err = msgFactory.NewMessageClient(types.MessageBusConfig{
+  PublishHost: types.HostInfo{
+  Host:     Configuration.MessageQueue.Host,
+  Port:     Configuration.MessageQueue.Port,
+  Protocol: Configuration.MessageQueue.Protocol,
+  },
+  Type: Configuration.MessageQueue.Type,})
+
+if err != nil {
+  LoggingClient.Error("failed to create messaging client: " + err.Error())
+}
+
+err = messsageBus.Connect()
+
+if err != nil {
+  LoggingClient.Error("failed to connect to message bus: " + err.Error())
+}
+```
+
+This code snippet shows how to publish a message to the abstract message bus.
+
+```go
+...
+payload, err := json.Marshal(evt)
+...
+msgEnvelope := types.MessageEnvelope{
+  CorrelationID: evt.CorrelationId,
+  Payload:       payload,
+  ContentType:   clients.ContentJson,
+}
+
+err = messageBus.Publish(msgEnvelope, Configuration.MessageQueue.Topic)
+```
+
+This code snippet shows how to subscribe to the abstract message bus.
+
+```go
+messageBus, err := factory.NewMessageClient(types.MessageBusConfig{
+  SubscribeHost: types.HostInfo{
+  Host:     Configuration.MessageQueue.Host,
+  Port:     Configuration.MessageQueue.Port,
+  Protocol: Configuration.MessageQueue.Protocol,
+  },
+  Type: Configuration.MessageQueue.Type,
+})
+
+if err != nil {
+  LoggingClient.Error("failed to create messaging client: " + err.Error())
+return
+}
+
+if err := messageBus.Connect(); err != nil {
+  LoggingClient.Error("failed to connect to message bus: " + err.Error())
+  return
+}
+
+topics := []types.TopicChannel{
+    {
+      Topic:    Configuration.MessageQueue.Topic,
+      Messages: messages,
+    },
+}
```

Comment:
In general I'm not clear from these docs how the topics are supposed to work. The above array would indicate that more than one can be specified. Is that true? How would I specify that in the configuration? How would I subscribe to multiple topics --say, JsonEvent and CborEvent? If I have two topics and I want to assign a different behavior when a message is received on a specific topic, what's the recommended practice? I assume I should create a subscriber for each topic. In the message consumption logic below, I don't see anything related to Topic other than logging statements, so I'm unclear on its usage.

Reply:
@tsconn23, yes it allows for multiple topics to be subscribed to with one call to subscribe. Specifying it in the service configuration depends on that services needs. Could be a map or just two fields in its Messaging struct. This library isn't dictating that.

The Topics have an associated channel that the messages for each topic are posted on. The service then does a select on each of it's topic channels and handles that topic message behavior.
A `subscriber` for each topic is not need. Just multiple TopicChannel items in the topics slice passed to Subscribe().

The example here is just specifying one topic and hence the one messages channel.
`case msgEnvelope := <-messages`

We took this approach for subscribing to multiple topics rather than requiring the user to call Subscribe() multiple times. If you prefer we could eliminate the slice and each call to Subscribe() is for just one topic. i.e the interface would be:
`Subscribe(topic string, messages chan *MessageEnvelope, errors chan error) error`